akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -367,11 +428,81 @@ object ByteString {
 
     def decodeString(charset: String): String = compact.decodeString(charset)
 
+    def decodeString(charset: Charset): String =
+      compact.decodeString(charset)
+
     private[akka] def writeToOutputStream(os: ObjectOutputStream): Unit = {
       os.writeInt(bytestrings.length)
       bytestrings.foreach(_.writeToOutputStream(os))
     }
 
+    override def take(n: Int): ByteString = {
+      @tailrec def take0(n: Int, b: ByteStringBuilder, bs: Vector[ByteString1]): ByteString =
+        if (bs.isEmpty || n <= 0) b.result
+        else if (n <= bs.head.length) b.append(bs.head.take(n)).result
+        else take0(n - bs.head.length, b.append(bs.head), bs.tail)
+
+      take0(n, ByteString.newBuilder, bytestrings)
```

Comment:
This allocates a new builder even if `n` <= 0 or `n` is as large, or larger, as `length`?